lib/sinon/test.js

```diff
@@ -55,6 +55,15 @@
                     sandbox.verifyAndRestore();
                 }
 
+                if (result && typeof result.then === "function" && typeof result.catch === "function") {
```

Comment:
It also feels like this should be an `else-if` of the previous `if` statement. Looking at the code now, a synchronous test that returns a promise will have restore called multiple times.

I'd add new tests to ensure that that is not the case.


If you move your if statement before the `typeof OldDone` check, you can actually make it so returning a promise is all that is necessary to make tests async instead of return a promise and taking a callback.

These are just my thoughts from looking at the code. I looked at issue this is fixing #1119 and that sample code looks synchronous but returns a promise. This doesn't yet fix that case. You now have a weird case where you're double-y saying it's synchronous.

Imagine this code:

```js
sinon.test(function(callback) {
  callback();
  // do lots of work here
  return Promise.reject(new Error);
});
```

I feel like we're enabling a bad situation the more I look at it. Mocha just [fixed this issue](/mochajs/mocha/blob/master/CHANGELOG.md#boom-breaking-changes) (see #4 on the list). I believe we should enforce the same restrictions. 

The end user's test should fall into one of these categories: 

* synchronous
* synchronous but returns a promise
* asynchronous